content/blog/eng/from-0-to-infinite-the-final-keyword.md

```diff
@@ -0,0 +1,338 @@
+
+The more precise definition:
+>“if S is a subtype of T, then objects of type T in a program may be replaced with objects of type S without altering any of >the desirable properties of that program.”
+
+Link to orginal paper: https://dl.acm.org/doi/pdf/10.1145/62138.62141
+
+> Simplifying the following could be good guidelines:
+
+> - Don’t stricter more validation rules on input parameters than those implemented by the parent class.
+> - Objects should be replaceable with instances of their subtypes without altering the correctness of that program
+> - Not add restriction in an overridden method if that restriction doesn't exist in baseline implementation
+
+I have used it like as example because is specifically for the design of hierarchies.
+
+I do not want to write an article on how to respect the development according to **Liskov** in the net there is a lot of literature. I would just like to emphasize that building a hierarchy of objects is not easy, that more the hierarchy is deep then more we are faced with design problems and **final** helps use to prevent this. 
+
+For this reason, before creating subclasses we should ask ourselves questions such as these or similar:
+
+- is my derived class of the same type as the base class?
+- I can use the derived class of base where requireds a base class without having "strange" behaviors in the code at runtime?
+- is my base class prepared to be derived?
+- the enhancements made by the subclass are primarily additive?
+
+If the answers to these questions are negative, using inheritance in this use case it could lead us to create complex and unmanageable hierarchies. 
+
+Now some basic example of wrong use of inheritance:
+```php
+use App\Domain\Model;
+class Stack extends ArrayList {
+    public push(Object value) { … }
+    public Object pop() { … }
+}
+```
+
+Why ? The Stack class inheriting from ArrayList will have a lot of additional methods not related to the concept of ArrayList (e.g. push and pop).
+
+In the previous example two main errors have been made, first error a Stack IS NOT an Array, second problem, is the cross-domain inheritance relationship. Simply, our domain classes must use implementations not to inherit them, in our example, stack was a concept of domain (Focusing only on reusing code can be a problem).
+
+You are probably thinking that the OOP is not keeping its promise. 
+However, let us remember that we are programming in OOP (Object Oriented Programming) not in OOI (Object Oriented Inheritance). There is always a trade off between code reuse and good design. 
+Inheritance should be mostly used for good design, for code reuse we go with composition.
+
+I invite you, however, to do some tests with your classes, to see if they adhere to these good practices and try to get a better design, avoiding inheritance.
```

Comment:
I invite you, however, to do some tests with your classes to see if they adhere to these good practices and try to get a better design by avoiding inheritance.